### DISCUSSION TITLE
Please change the free license to LGPL

### DISCUSSION THREAD

**Discussion by TheQuantumPhysicist** [slint-ui/slint#379] [General]
I'm new to this library, and it seems like it has good ideas. The most important factor for this library to grow is to have a license that can be freely used even in Linux for everyone. With GPL, this library will be feared because everyone will have to make their code open source. Especially that GPL "infects" all other libraries used in a project and forces them to be open source. However, you can still get all the benefits of the commercial license even with LGPL. Please consider that.

**Comment by joseluis** [comment]
Yeah, the fact that anything that you create that links [even dynamically](https://www.gnu.org/licenses/gpl-faq.en.html#GPLStaticVsDynamic) to this library must also be licensed as GPL-3 is a deal breaker for me. Specially coming from the Rust ecosystem where 99% is dual licensed Apache 2 + MIT. It's a shame because otherwise the project looks very promising.

**Reply by bjorn** [reply]
> anything that you create that links [even dynamically](https://www.gnu.org/licenses/gpl-faq.en.html#GPLStaticVsDynamic) to this library must also be licensed as GPL-3 is a deal breaker for me

You can release your software under Apache 2 and/or MIT as well since they are GPL-3 compatible. Only if you want to release a proprietary product you need a commercial license for SixtyFPS.

**Reply by joseluis** [reply]
> You can release your software under Apache 2 and/or MIT as well since they are GPL-3 compatible. Only if you want to release a proprietary product you need a commercial license for SixtyFPS.

Oh, I wasn't aware of that. thank you. That settles it then

**Reply by martynp** [reply]
> You can release your software under Apache 2 and/or MIT as well since they are GPL-3 compatible. Only if you want to release a proprietary product you need a commercial license for SixtyFPS.

That isn't true though.

Compatibility only works for GPLv3 licensed works - so a GPLv3 work can incorporate Apache 2 and MIT licensed works, but an Apache 2 work cannot contain a GPLv3.

So Slint can use Apache 2 and MIT libraries, but if your product is Apache 2 and you have dependency on Slint then your product must become GPLv3 when distributed or it is not compatible.

I am not even sure LGPLv3 would be any better - Rust doesn't play well with the concept of linking. The intention of the LGPL is that an end user can replace the LGPL portion of a product with their own one (assuming linking compatibility) - the way Rust is compiled it isn't possible to do that at a binary level.

(Sorry for digging up an old discussion, but I was weighing pro's and con's of using Slint and GPL is at least a pain as I would want to release Apache 2 to avoid the ongoing contractual aspects of GPLv3 in commercial products)

**Reply by ogoffart** [reply]
> if your product is Apache 2 and you have dependency on Slint then your product must become GPLv3 when distributed or it is not compatible.

That's correct. The final product needs to be GPLv3. 
But you can still license your own code for that product under the Apache 2 since that is compatible.

> to avoid the ongoing contractual aspects of GPLv3 in commercial products

For commercial products (That cannot use the GPL), you simply need to get a commercial license from us. The [ambassador license](https://slint-ui.com/ambassador-program.html)  is for free.

**Reply by martynp** [reply]
I am not sure how I could license it as Apache 2 if it uses Slint as the two are incompatible in that direction. If my code has a dependency on GPLv3 code, there is no way I can license it as anything else without completely removing a dependency on GPLv3 code.

Sure I own the copyright, but as long as it is derivative of GPLv3 then that sticks.

The software aspect of the commercial product is Apache 2, so I could not use Slint without moving the rest of the code connected with the UI to being GPLv3 or close source the project.

**Reply by ogoffart** [reply]
Yes, the final product need to be GPLv3. 
Having your code as Apache would allow a downstream user to either release its product as GPLv3, or chose to get a Slint license if they want to distribute the software under a proprietary license.

> I could not use Slint without moving the rest of the code connected with the UI to being GPLv3 or close source the project.

You can still keep your project open-souce and get a Slint license for for the proprietary part.

Please feel free to contact us per email if you want to discuss the specifics for your use case.

**Comment by ogoffart** [comment]
The license is GPL on purpose.
We hope to get revenue from SixtyFPS. The GPL is a good fit for open-soure users.
People who don't want to open-source their product, will be able to obtain a commercial license.
We haven't finalized yet the exact pricing model but hope to be something affordable.

**Comment by TheQuantumPhysicist** [comment]
You're mistaken, my man. Allow me to explain. You're supporting a small set of open source developers. Often times I see projects that have to keep parts of their code closed because of licensing issues. Right now, everyone who has any segment of their code with a license incompatible with GPL, will not be able to use your library. Notice: They **won't** be able to use your library. 

Here's why GPL is useless for you to make money: Any company that will use your code, and keeps it internal, they still have the right to use your code without making the code open source. On the other hand, if any company uses your library and want to make users benefit from it, they will anyway be unable to do that because they can't make money out of your library, because your free license is restricted to "non-commercial use". After all, all you want is to prevent people making money without you being part of it, right?

I often recommend to library maintainers to think of it this way: "What is it that you're afraid to happen?" You want more people to use your library. You want even more companies to use your library and get "hooked" with the beauty of your library, per se, which will give them an incentive to consider buying a license. Often companies that I worked for, anyone who saw "GPL" will just dismiss that library. No questions asked. Do you really want that? People often start using libraries for free, then they pay for them when they succeed. All those people won't use your library because of GPL. I bet you'll lose more money than you'll make by restricting it.

It's up to you. Consider my view in your discussions. After all, no company is gonna make money out of your library without paying you if you clearly put it in the license as "non-commercial".

Best of luck.

**Reply by bjorn** [reply]
> they can't make money out of your library, because your free license is restricted to "non-commercial use"

I think it's important to point out that the GPL is actually not a non-commercial license. It wouldn't be a free software license if it would restrict usage to non-commercial use.

**Comment by tronical** [comment]
> It's up to you. Consider my view in your discussions. After all, no company is gonna make money out of your library without paying you if you clearly put it in the license as "non-commercial".

Thank you for your input - it is appreciated. In our past careers we've had a very positive experience with the model that we've chosen for this project.

> Best of luck.

Thank you!

**Comment by ogoffart** [comment]
I moved the issue to a GitHub discussion.

> "What is it that you're afraid to happen?"

Well, we're afraid that some users use our library for free while we think they should pay.
With the LGPL, any entity can use the library for free (almost) without conditions, and we don't want that. Maybe in the future we could consider if that would be an actual business model, but since we can hardly go back to GPL, this is not an option for us at this point.

With the GPL, we are happy to let open source users use the library for free.
Commercial users still will be able to use the library because of the escape hatch of the dual license.

We want that the commercial license fits your use case so the most people can still use the library. We're even considering a free (as in beer) license for the early-adopters and "ambassadors".
What's the concrete use case you have in mind  that will prevent to use the library? We can try to accommodate this case.

**Reply by pinobatch** [reply]
>  What's the concrete use case you have in mind that will prevent to use the library? We can try to accommodate this case.

Any application for iPhone or iPad, once Slint supports those devices. From README.md:

> Mobile (Android/iOS): Todo. We haven't started supporting mobile platforms yet, but it is our intention to do so in the future.

Some computing platforms are configured to use only one application distribution service, which imposes conditions incompatible with the GNU GPL. These services include Apple's App Store and the download stores of the major video game consoles. Someone distributing an application using the planned iOS port of the library would not be allowed to use the GNU GPL version. Thus all users making iOS applications would have to be "ambassadors" if the application is distributed without charge, as I can't think of a way to recover the [current Distribution License fee of 1 EUR per download](https://slint-ui.com/pricing.html#distribution-license) from a free (as in beer) application.

**Reply by ogoffart** [reply]
> Thus all users making iOS applications would have to be "ambassadors" if the application is distributed without charge

That sounds about right. However I still don't see why this is a problem as the ambassador license is free of charge and fairly easy to follow.
Maybe we'll consider something for the iOS case when we do the port.

> I can't think of a way to recover the [current Distribution License fee of 1 EUR per download](https://slint-ui.com/pricing.html#distribution-license) from a free (as in beer) application.

Maybe that's not 100% clear in our explanation, but the 1€ fee is not per downloads, it is per hardware device sold with slint in it.
